firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java

```diff
@@ -61,19 +61,28 @@ public boolean isEnabled() {
     return dataCollectionDefaultEnabled.get();
   }
 
-  public void setEnabled(boolean enabled) {
-    if (dataCollectionDefaultEnabled.compareAndSet(!enabled, enabled)) {
-      sharedPreferences.edit().putBoolean(DATA_COLLECTION_DEFAULT_ENABLED, enabled).apply();
-
-      publisher.publish(
-          new Event<>(DataCollectionDefaultChange.class, new DataCollectionDefaultChange(enabled)));
+  public void setEnabled(Boolean enabled) {
+    if (enabled == null) {
+      sharedPreferences.edit().remove(DATA_COLLECTION_DEFAULT_ENABLED).apply();
```

Comment:
Should this also live inside the synchronized block?

Reply:
I think shared preferences itself makes sure that it's synchornized

Reply:
Sure, you may however run into a situation where one thread sets the shared prefs value and another one broadcasts the message. So they need to be synchronized around the same mutex

Reply:
I dont think a synchronized block is needed in this case. Since the method is doing only one write and no read from the shared prefs. And since readManifestDataCollectionEnabled always returns the same value based on the manifest file

Reply:
You *do* need to synchronize here, before your change synchronization was happening with the atomic bool, but since you are checking for null now, a bunch of races become possible, i.e. concurrent threads can enter different branches of this if block and cause mess up the resulting state.

Reply:
actually nvm the setEnabled method should be in a sync block..
setEnabled(true) -> Thread1
setEnabled(false) -> Thread2
This might result in the shared preferences having the value of false but the dataCollection having the value of True based on the order of execution